netbox/dcim/resource_netbox_dcim_rack_test.go

```diff
@@ -0,0 +1,159 @@
+	})
+}
+
+func TestAccNetboxDcimRackFull(t *testing.T) {
+	nameSuffix := acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { util.TestAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckNetboxDcimRackConfig(nameSuffix, true, true),
+				Check: resource.ComposeTestCheckFunc(
+					util.TestAccResourceExists(resourceNameNetboxDcimRack),
+				),
+			},
+			{
+				ResourceName:      resourceNameNetboxDcimRack,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccNetboxDcimRackMinimalFullMinimal(t *testing.T) {
+	nameSuffix := acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { util.TestAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckNetboxDcimRackConfig(nameSuffix, false, false),
+				Check: resource.ComposeTestCheckFunc(
+					util.TestAccResourceExists(resourceNameNetboxDcimRack),
+				),
+			},
+			{
+				Config: testAccCheckNetboxDcimRackConfig(nameSuffix, true, true),
+				Check: resource.ComposeTestCheckFunc(
+					util.TestAccResourceExists(resourceNameNetboxDcimRack),
+				),
+			},
+			{
+				Config: testAccCheckNetboxDcimRackConfig(nameSuffix, false, true),
+				Check: resource.ComposeTestCheckFunc(
+					util.TestAccResourceExists(resourceNameNetboxDcimRack),
+				),
+			},
+			{
+				Config: testAccCheckNetboxDcimRackConfig(nameSuffix, false, false),
+				Check: resource.ComposeTestCheckFunc(
+					util.TestAccResourceExists(resourceNameNetboxDcimRack),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckNetboxDcimRackConfig(nameSuffix string, resourceFull, extraResources bool) string {
+	template := `
+  {{ if eq .extraresources "true" }}
+  resource "netbox_extras_tag" "test" {
+		name = "test-{{ .namesuffix }}"
+		slug = "test-{{ .namesuffix }}"
+	}
+
+  resource "netbox_dcim_rack_role" "test" {
+    name = "test-{{ .namesuffix }}"
+    slug = "test-{{ .namesuffix }}"
+  }
+
+  resource "netbox_tenancy_tenant" "test" {
+    name = "test-{{ .namesuffix }}"
+    slug = "test-{{ .namesuffix }}"
+  }
+
+  resource "netbox_dcim_location" "test" {
+		name        = "test-{{ .namesuffix }}"
+    site_id     = netbox_dcim_site.test.id
+		slug        = "test-{{ .namesuffix }}"
+  }
+	{{ end }}
+
+  resource "netbox_dcim_site" "test" {
+    name = "test-{{ .namesuffix }}"
+    slug = "test-{{ .namesuffix }}"
+  }
+
+	resource "netbox_dcim_rack" "test" {
+		name        = "test-{{ .namesuffix }}"
+    site_id     = netbox_dcim_site.test.id
+    height      = 10
+    width       = 10
+
+    {{ if eq .resourcefull "true" }}
+		asset_tag = "test-{{ .namesuffix }}"
+    comments = <<-EOT
+    Comments for Test Rack
+    Multiline
+    EOT
+    desc_units = true
+    facility = "test-{{ .namesuffix }}"
+    // location_id = netbox_dcim_location.test.id
```

Comment:
Hello @amhn,

Do you have time to see why I am not able to test this ?  
A bug with Netbox ?

Thanks.

Reply:
If I activate line 137 and run the test, tcpdump shows the following
```
{"location":["Related object not found using the provided numeric ID: 0"]}
```

Reply:
```go
    diff --git a/netbox/dcim/resource_netbox_dcim_rack.go b/netbox/dcim/resource_netbox_dcim_rack.go
index 0cbc8ffc..8d6cef61 100644
--- a/netbox/dcim/resource_netbox_dcim_rack.go
+++ b/netbox/dcim/resource_netbox_dcim_rack.go
@@ -378,7 +378,7 @@ func resourceNetboxDcimRackUpdate(ctx context.Context, d *schema.ResourceData,
        if d.HasChange("location_id") {
                locationID := int64(d.Get("location_id").(int))
                params.Location = &locationID
-               modifiedFields["group"] = locationID
+               modifiedFields["location"] = locationID
        }
        if d.HasChange("name") {
                name := d.Get("name").(string)
```

This seems to fix it. Need to look into it further.

The failure is in step 3. This step unassigns location from rack. New location is 0 (because for terrafom 0 == null). This results in `"location": 0` in the json. Request-modifier needs to overwrite the field with null, so that netbox will accept it.